lib/Mongo/MongoWriteBatch.php

```diff
@@ -116,54 +119,62 @@ public function add($item)
         }
 
         $collection = $this->collection->getCollection();
+        $exception = null;
 
         try {
-            $result = $collection->BulkWrite($this->items, $options);
+            $writeResult = $collection->bulkWrite($this->items, $options);
+            $resultDocument = [];
             $ok = true;
-        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
-            $result = $e->getWriteResult();
+        } catch (BulkWriteException $e) {
+            $writeResult = $e->getWriteResult();
+            $resultDocument = ['writeErrors' => $this->convertWriteErrors($writeResult)];
             $ok = false;
         }
 
-        if ($ok === true) {
-            $this->items = [];
-        }
+        $this->items = [];
 
         switch ($this->batchType) {
             case self::COMMAND_UPDATE:
                 $upsertedIds = [];
-                foreach ($result->getUpsertedIds() as $index => $id) {
+                foreach ($writeResult->getUpsertedIds() as $index => $id) {
                     $upsertedIds[] = [
                         'index' => $index,
                         '_id' => TypeConverter::toLegacy($id)
                     ];
                 }
 
-                $result = [
-                    'nMatched' => $result->getMatchedCount(),
-                    'nModified' => $result->getModifiedCount(),
-                    'nUpserted' => $result->getUpsertedCount(),
-                    'ok' => $ok,
+                $resultDocument += [
+                    'nMatched' => $writeResult->getMatchedCount(),
+                    'nModified' => $writeResult->getModifiedCount(),
+                    'nUpserted' => $writeResult->getUpsertedCount(),
+                    'ok' => true,
                 ];
 
                 if (count($upsertedIds)) {
-                    $result['upserted'] = $upsertedIds;
+                    $resultDocument['upserted'] = $upsertedIds;
                 }
-
-                return $result;
+                break;
 
             case self::COMMAND_DELETE:
-                return [
-                    'nRemoved' => $result->getDeletedCount(),
-                    'ok' => $ok,
+                $resultDocument += [
+                    'nRemoved' => $writeResult->getDeletedCount(),
+                    'ok' => true,
                 ];
+                break;
 
             case self::COMMAND_INSERT:
-                return [
-                    'nInserted' => $result->getInsertedCount(),
-                    'ok' => $ok,
+                $resultDocument += [
+                    'nInserted' => $writeResult->getInsertedCount(),
+                    'ok' => true,
                 ];
+                break;
+        }
+
+        if (! $ok) {
+            throw new \MongoWriteConcernException('Failed write', 911, null, $resultDocument);
```

Comment:
> I would not return the error code hardcoded, but use the error code of the previous MongoWriteConcernException.

Yeah, that was the quick hack to ensure it's passing, I'll write more tests adding failures during update and remove batches to ensure proper exception codes are passed along.

> Also there is a \MongoDuplicateKeyException, would be nice if I would get that specific exception class from this lib, too.

The check actually produces an error due to a unique key violation but ext-mongo still throws `\MongoWriteConcernException`. I'm not sure where you'd expect `\MongoDuplicateKeyException`.